frontend/lib/resolution/try-catch-analysis.cpp

```diff
@@ -0,0 +1,364 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "try-catch-analysis.h"
+
+#include "chpl/parsing/parsing-queries.h"
+#include "chpl/resolution/ResolvedVisitor.h"
+#include "chpl/resolution/resolution-types.h"
+#include "chpl/resolution/scope-queries.h"
+#include "chpl/uast/all-uast.h"
+#include "chpl/framework/compiler-configuration.h"
+#include "chpl/resolution/can-pass.h"
+#include "Resolver.h"
+
+
+namespace chpl {
+namespace resolution {
+
+
+using namespace uast;
+using namespace types;
+
+
+enum class ErrorCheckingMode {
+  UNKNOWN,
+  FATAL,
+  RELAXED,
+};
+
+  struct TryCatchState {
+    const Try* tryNode;
+    int taskFunctionDepth;
+    bool catchesAreExhaustive;
+  };
+
+  struct TryCatchAnalyzer {
+    using RV = ResolvedVisitor<TryCatchAnalyzer>;
+
+    // input
+    Context* context;
+    const AstNode* astForErr;
+    ErrorCheckingMode mode;
+    // need to keep a stack of try nodes and some meta info
+    std::vector<TryCatchState> tryStack;
+    const Function* currentFn = nullptr;
+    // output
+    // TODO: figure out the output - errors when something doesn't pass?
+    // maybe no output?
+    std::vector<QualifiedType> returnedTypes;
+
+    TryCatchAnalyzer(Context* context,
+                  const AstNode* symbol,
+                  ErrorCheckingMode mode)
+        : context(context),
+          astForErr(symbol),
+          mode(mode) {
+    }
+
+    void checkCatchallIsLast(const Try* node, RV& rv) {
+      bool hasCatchAll = false;
+      for (auto handler : node->handlers()) {
+        if (hasCatchAll) {
+          context->error(node, "catchall placed before the end of a catch list");
+        }
+        if (isCatchAll(handler, rv)) {
+          hasCatchAll = true;
+        }
+      }
+    }
+
+    // Returns true if the catch is a catchall statement.
+    bool isCatchAll(const Catch* node, RV& rv) {
+        bool catchAll = false;
+        if (auto err = node->error()) {
+          // we validate that the type of error is a subtype of Error or Error
+          // itself resolving the AST.
+          // So here we just need to check for the base Errors.Error to know if
+          // this is a catchall.
+          const ResolvedExpression& r = rv.byAst(err);
+          auto errType = CompositeType::getErrorType(context);
+          auto dec = ClassTypeDecorator(ClassTypeDecorator::MANAGED_NONNIL);
+          auto manager = AnyOwnedType::get(context);
+          auto ret = ClassType::get(context, errType->basicClassType(),
+                                    manager, dec);
+          auto qt = QualifiedType(QualifiedType::VAR,
+                                  ret->withDecorator(context, dec));
+          auto passResult = canPass(context, r.type(), qt);
+          using CPR=CanPassResult::ConversionKind;
+          if (passResult.passes() && passResult.conversionKind()==CPR::NONE) {
```

Comment:
At the risk of seeming very nitpicky... 
```suggestion
          using CPR = CanPassResult::ConversionKind;
          if (passResult.passes() && passResult.conversionKind() == CPR::NONE) {
```